libshvcore/src/utils/getlog.cpp

```diff
@@ -0,0 +1,291 @@
+
+#include <shv/core/log.h>
+#include <shv/core/utils/abstractshvjournal.h>
+#include <shv/core/utils/getlog.h>
+#include <shv/core/utils/patternmatcher.h>
+#include <shv/core/utils/shvlogheader.h>
+
+#include <functional>
+#include <ranges>
+
+#define logWGetLog() shvCWarning("GetLog")
+#define logIGetLog() shvCInfo("GetLog")
+#define logMGetLog() shvCMessage("GetLog")
+#define logDGetLog() shvCDebug("GetLog")
+
+namespace shv::core::utils {
+using chainpack::RpcValue;
+
+namespace {
+[[nodiscard]] RpcValue as_shared_path(chainpack::RpcValue::Map& path_map, const std::string &path)
+{
+	RpcValue ret;
+	if (auto it = path_map.find(path); it == path_map.end()) {
+		ret = static_cast<int>(path_map.size());
+		logDGetLog() << "Adding record to path cache:" << path << "-->" << ret.toCpon();
+		path_map[path] = ret;
+	} else {
+		ret = it->second;
+	}
+	return ret;
+}
+
+struct GetLogContext {
+	RpcValue::Map pathCache;
+	ShvGetLogParams params;
+};
+
+enum class Status {
+	Ok,
+	RecordCountLimitHit
+};
+
+void append_log_entry(RpcValue::List& log, const ShvJournalEntry &e, GetLogContext& ctx)
+{
+	logDGetLog() << "\t append_log_entry:" << e.toRpcValueMap().toCpon();
+	log.push_back(RpcValue::List{
```

Comment:
we have `ShvJournalEntry::toRpcValueList()` already

Reply:
Done